### PR TITLE
fix(list, combobox, tree): fix EvalError with CSP on renderer

### PR DESCRIPTION
**Pull request by Theeraphat-Sorasetsakul**
## Description

Uncaught EvalError deploying apps with Content-Security-Policy, affecting ef-combo-box and potentially list-related elements.

**Reproduce steps:**

1. Add ef-combo-box to index.html.
2. Apply <meta http-equiv="Content-Security-Policy"content="script-src 'self' [https://*.refinitiv.com](https://%2A.refinitiv.com/)">
3. Observe Uncaught EvalError.

Fixes # ([DME-2088](https://jira.refinitiv.com/browse/DME-2088))

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 1]
```diff
@@ -1,30 +1,43 @@
...
+import type { CollectionComposer, DataItem } from '@refinitiv-ui/utils/collection.js';
 
 import type { Item, ItemData } from '../../item';
-import { ListRenderer } from '../../list/index.js';
+import { createListRenderer } from '../../list/index.js';
 import { Renderer } from '../../list/renderer.js';
 
+export const createComboBoxRenderer = <T extends DataItem = ItemData>(
+  context?: unknown
+): ((item: T, composer: CollectionComposer<T>, element?: HTMLElement) => HTMLElement) => {
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]: This looks unnecessary.
```suggestion
) => {
```

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 2]
```diff
@@ -1,30 +1,43 @@
-import type { CollectionComposer } from '@refinitiv-ui/utils/collection.js';
+import { DeprecationNotice } from '@refinitiv-ui/core';
+
+import type { CollectionComposer, DataItem } from '@refinitiv-ui/utils/collection.js';
 
 import type { Item, ItemData } from '../../item';
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]: unused `Item`

File: packages/elements/src/combo-box/index.ts [hunk 3]
```diff
@@ -119,10 +119,10 @@ export class ComboBox<T extends DataItem = ItemData> extends FormFieldElement {
 
   /**
    * Renderer used to render list item elements
-   * @type {ComboBoxRenderer}
+   * @type {createComboBoxRenderer}
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/index.ts]: let's remove this `@type` as it's duplicate with type info from TypeScript itself.

File: packages/elements/src/list/elements/list.ts [hunk 4]
```diff
@@ -85,10 +85,10 @@ export class List<T extends DataItem = ItemData> extends ControlElement {
 
   /**
    * Renderer used to render list item elements
-   * @type {ListRenderer}
+   * @type {createListRenderer}
```
Comment (wattachai-lseg) [inline packages/elements/src/list/elements/list.ts]: let's remove this @type as it's duplcate with type info from TypeScript itself.

File: packages/elements/src/list/helpers/renderer.ts [hunk 5]
```diff
@@ -12,47 +14,57 @@ type Context = {
   multiple?: boolean;
 };
 
+export const createListRenderer = <T extends DataItem = ItemData>(
+  context?: unknown
+): ((item: T, composer: CollectionComposer<T>, element?: HTMLElement) => HTMLElement) => {
```
Comment (wattachai-lseg) [inline packages/elements/src/list/helpers/renderer.ts]: This looks unnecessary.
```suggestion
) => {
```

File: packages/elements/src/tree/elements/tree.ts [hunk 6]
```diff
@@ -72,7 +72,7 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
    * @type {TreeRenderer}
```
Comment (wattachai-lseg) [inline packages/elements/src/tree/elements/tree.ts]: let's remove this `@type` as it's duplicate with type info from TypeScript itself.

File: packages/elements/src/tree/elements/tree.ts [hunk 7]
```diff
@@ -7,7 +7,7 @@ import { CollectionComposer } from '@refinitiv-ui/utils/collection.js';
 import { List, valueFormatWarning } from '../../list/index.js';
 import { VERSION } from '../../version.js';
 import { defaultFilter } from '../helpers/filter.js';
-import { TreeRenderer } from '../helpers/renderer.js';
+import { TreeRenderer, createTreeRenderer } from '../helpers/renderer.js';
```
Comment (wattachai-lseg) [inline packages/elements/src/tree/elements/tree.ts]: unused `TreeRenderer`

File: packages/elements/src/tree/helpers/renderer.ts [hunk 8]
```diff
@@ -11,49 +14,59 @@ type RendererScope = {
   noRelation?: boolean;
 };
 
+export const createTreeRenderer = <T extends TreeDataItem = TreeDataItem>(
+  scope?: unknown
+): ((item: T, composer: CollectionComposer<T>, element?: HTMLElement) => HTMLElement) => {
```
Comment (wattachai-lseg) [inline packages/elements/src/tree/helpers/renderer.ts]: This looks unnecessary.
```suggestion
) => {
```

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 9]
```diff
@@ -1,30 +1,43 @@
...
-import { ListRenderer } from '../../list/index.js';
+import { createListRenderer } from '../../list/index.js';
 import { Renderer } from '../../list/renderer.js';
 
+export const createComboBoxRenderer = <T extends DataItem = ItemData>(
+  context?: unknown
+): ((item: T, composer: CollectionComposer<T>, element?: HTMLElement) => HTMLElement) => {
+  const listRenderer = createListRenderer<T>(context);
+
+  return (item: T, composer: CollectionComposer<T>, element?: HTMLElement) => {
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]:
```suggestion
  return (item: T, composer: CollectionComposer<T>, element?: HTMLElement): HTMLElement => {
```

File: packages/elements/src/tree-select/index.ts [hunk 10]
```diff
@@ -239,7 +242,7 @@ export class TreeSelect extends ComboBox<TreeSelectDataItem> {
    * @type {TreeSelectRenderer}
```
Comment (wattachai-lseg) [inline packages/elements/src/tree-select/index.ts]: let's remove this @type as it's duplicate with type info from TypeScript itself.

**Comment by wattachai-lseg** [comment]
**Update: All resolved. 🎉**

`/packages/elements/src/list/renderer.ts`
* `RenderFunction` & `RendererConstructor` in should be removed as part of this PR.
* `Renderer` should be marked as deprecated with a comment that it should be removed along with the Renderer classes.

`/packages/elements/src/list/extensible-function.ts`
* `ExtensibleFunction` should be marked as deprecated with a comment that it should removed along with the Renderer classes.

**Default Renderer Usage**
* `createRenderer()` should be called with type param.

File: packages/elements/src/combo-box/helpers/renderer.ts (hunk 2 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/combo-box/helpers/renderer.ts]:
```suggestion
import type { ItemData } from '../../item';
```

File: packages/elements/src/combo-box/index.ts (hunk 3 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/combo-box/index.ts]:
```suggestion
```

File: packages/elements/src/list/elements/list.ts (hunk 4 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/list/elements/list.ts]:
```suggestion
```

File: packages/elements/src/tree/elements/tree.ts (hunk 6 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/tree/elements/tree.ts]:
```suggestion
```

File: packages/elements/src/tree/elements/tree.ts (hunk 7 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/tree/elements/tree.ts]:
```suggestion
import { createTreeRenderer } from '../helpers/renderer.js';
```

File: packages/elements/src/tree-select/index.ts (hunk 10 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/tree-select/index.ts]:
```suggestion
```

File: packages/elements/src/tree/helpers/renderer.ts [hunk 11]
```diff
@@ -11,49 +14,59 @@ type RendererScope = {
...
+   */
+  const key: string = uuid();
+
+  let manager: TreeManager<T>;
+  let currentMode: TreeManagerMode;
+  let currentComposer: CollectionComposer<T>;
+  return (
+    item: T,
+    composer: CollectionComposer<T>,
+    element = document.createElement('ef-tree-item')
```
Comment (wattachai-lseg) [inline packages/elements/src/tree/helpers/renderer.ts]:
```suggestion
    element: HTMLElement = document.createElement('ef-tree-item')
```

File: packages/elements/src/tree-select/index.ts [hunk 12]
```diff
@@ -27,13 +27,16 @@ import type { Overlay } from '../overlay';
 import type { Pill } from '../pill';
 import '../pill/index.js';
 import '../tree/index.js';
```
Comment (wattachai-lseg) [inline packages/elements/src/tree-select/index.ts]: I don't think we need this line.

Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/tree-select/index.ts]: Removed.

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 13]
```diff
@@ -1,30 +1,42 @@
...
+};
+
+const deprecationNotice = new DeprecationNotice(
+  'ComboRenderer is deprecated. Use createComboBoxRenderer instead.'
+);
+
 /**
  * Renders list items as `ef-item` elements.
  * Extends its behaviour from ListRenderer.
+ * @deprecate Don't use this! It causes uncaught EvalError when using script-src 'self' CSP.
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]: Good job! We should add deprecation to jsdoc too. However, the message should be the same with deprecation notice.
```suggestion
 * @deprecate ComboRenderer is deprecated. Use createComboBoxRenderer instead.
```

File: packages/elements/src/list/extensible-function.ts [hunk 14]
```diff
@@ -1,7 +1,6 @@
 /**
  * Extensible function class
- * TODO: Move this to @refinitiv-ui/utils
- * ! Do not import this module !
+ * @deprecate Don't use this! It causes uncaught EvalError when using script-src 'self' CSP.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/extensible-function.ts]:
```suggestion
 * @deprecate ExtensibleFunction is deprecated.
```

File: packages/elements/src/list/extensible-function.ts [hunk 15]
```diff
@@ -1,7 +1,6 @@
 /**
  * Extensible function class
```
Comment (wattachai-lseg) [inline packages/elements/src/list/extensible-function.ts]:
```suggestion
 * ExtensibleFunction is not CSP compliant. Do not use.
```

File: packages/elements/src/list/helpers/renderer.ts [hunk 16]
```diff
@@ -12,47 +14,56 @@ type Context = {
...
+};
+
+const deprecationNotice = new DeprecationNotice(
+  'ListRenderer is deprecated. Use createListRenderer instead.'
+);
+
 /**
  * Renders list items as `ef-item` elements.
  * This is the default renderer for lists.
+ * @deprecate Don't use this! It causes uncaught EvalError when using script-src 'self' CSP.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/helpers/renderer.ts]:
```suggestion
 * @deprecate ListRenderer is deprecated. Use createListRenderer instead.
```

File: packages/elements/src/list/renderer.ts [hunk 17]
```diff
@@ -1,39 +1,8 @@
...
-   */
-  new (fn: RenderFunction): RenderFunction;
-}
-
 /**
  * Renderer base class.
  * Used for creating renderers to render data items.
- * TODO: Move this to @refinitiv-ui/utils
- * ! Do not import this module !
+ * @deprecate Don't use this! It causes uncaught EvalError when using script-src 'self' CSP.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/renderer.ts]:
```suggestion
 * @deprecate Renderer is deprecated.
```

File: packages/elements/src/list/renderer.ts [hunk 18]
```diff
@@ -1,39 +1,8 @@
...
- */
-export interface RendererConstructor {
-  /**
-   * @param fn Render function to use as the instance
-   */
-  new (fn: RenderFunction): RenderFunction;
-}
-
 /**
  * Renderer base class.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/renderer.ts]:
```suggestion
 * Renderer is not CSP compliant. Do not use.
```

File: packages/elements/src/tree/helpers/renderer.ts [hunk 19]
```diff
@@ -11,49 +14,60 @@ type RendererScope = {
...
+    return _element;
+  };
+};
+
+const deprecationNotice = new DeprecationNotice(
+  'TreeRenderer is deprecated. Use createTreeRenderer instead.'
+);
+
+/**
+ * @deprecate Don't use this! It causes uncaught EvalError when using script-src 'self' CSP.
```
Comment (wattachai-lseg) [inline packages/elements/src/tree/helpers/renderer.ts]:
```suggestion
 * @deprecate TreeRenderer is deprecated. Use createTreeRenderer instead.
```

File: packages/elements/src/combo-box/__demo__/index.html [hunk 20]
```diff
@@ -203,49 +203,47 @@
...
         import '@refinitiv-ui/elements/flag';
 
-        class FlagRender extends ComboBoxRenderer {
-          constructor(context) {
-            // Keep the reference to the default renderer
-            const renderer = super(context);
-            // store reference to flag for easy access.
-            // Use WeakMap to not care about memory leaks
-            const flagMap = new WeakMap();
+        const flagRender = (context) => {
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/__demo__/index.html]:
```suggestion
        const createFlagRender = (context) => {
```

File: packages/elements/src/combo-box/__demo__/index.html [hunk 21]
```diff
@@ -203,49 +203,47 @@
...
-            };
-          }
-        }
+            return element;
+          };
+        };
 
         const customRenderer = document.getElementById('custom-renderer');
-        customRenderer.renderer = new FlagRender(customRenderer);
+        customRenderer.renderer = flagRender(customRenderer);
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/__demo__/index.html]:
```suggestion
        customRenderer.renderer = createFlagRender(customRenderer);
```

File: packages/elements/src/combo-box/__demo__/index.html
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/__demo__/index.html:1]: missing flags from custom render demo-block. Missing theme import for Flag component?

File: documents/src/pages/elements/combo-box.md [hunk 22]
```diff
@@ -451,17 +451,17 @@ comboBox.addEventListener('query-changed', (event) => {
 
 Combo Box supports custom rendering by providing a renderer function to the `renderer` property. The renderer receives a data item, Collection Composer and previously mapped item elements (if any), and must return an `HTMLElement`.
 
-The preferred approach is to create new renderer reference to the `ComboBoxRenderer` that comes with Combo Box. The default renderer uses [Item](./elements/item) elements, and supports highlighted, selected, disabled, hidden and readonly states.
+The preferred approach is to create new renderer reference to the `createComboBoxRenderer` that comes with Combo Box. The default renderer uses [Item](./elements/item) elements, and supports highlighted, selected, disabled, hidden and readonly states.
```
Comment (wattachai-lseg) [inline documents/src/pages/elements/combo-box.md]:
```suggestion
The preferred approach is to create new renderer reference with `createComboBoxRenderer` that comes with Combo Box. The default renderer uses [Item](./elements/item) elements, and supports highlighted, selected, disabled, hidden and readonly states.
```

File: packages/elements/src/combo-box/__demo__/index.html
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/combo-box/__demo__/index.html:1]: Added.

File: packages/elements/src/combo-box/__demo__/index.html [hunk 23]
```diff
@@ -203,49 +205,47 @@
     <demo-block header="Custom renderer" layout="normal" tags="renderer">
       <ef-combo-box id="custom-renderer"></ef-combo-box>
       <script type="module">
-        import { ComboBoxRenderer } from '@refinitiv-ui/elements/combo-box';
+        import { createComboBoxRenderer } from '@refinitiv-ui/elements/combo-box';
         import '@refinitiv-ui/elements/flag';
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/__demo__/index.html]: duplicate import?

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 24]
```diff
@@ -1,30 +1,42 @@
...
+};
+
+const deprecationNotice = new DeprecationNotice(
+  'ComboRenderer is deprecated. Use createComboBoxRenderer instead.'
+);
+
 /**
  * Renders list items as `ef-item` elements.
  * Extends its behaviour from ListRenderer.
+ * @deprecate ComboRenderer is deprecated. Use createComboBoxRenderer instead.
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]: Dang, I missed the typo. Could you also update the one in `deprecationNotice`?
```suggestion
 * @deprecate ComboBoxRenderer is deprecated. Use createComboBoxRenderer instead.
```

File: packages/elements/src/list/extensible-function.ts [hunk 25]
```diff
@@ -1,7 +1,6 @@
 /**
- * Extensible function class
- * TODO: Move this to @refinitiv-ui/utils
- * ! Do not import this module !
+ * ExtensibleFunction is not CSP compliant. Do not use.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/extensible-function.ts]: My bad. Action should come first.
```suggestion
 *  Do not use: ExtensibleFunction is not CSP compliant.
```

File: packages/elements/src/list/renderer.ts [hunk 26]
```diff
@@ -1,39 +1,8 @@
...
-export interface RendererConstructor {
-  /**
-   * @param fn Render function to use as the instance
-   */
-  new (fn: RenderFunction): RenderFunction;
-}
-
-/**
- * Renderer base class.
+ * Renderer is not CSP compliant. Do not use.
```
Comment (wattachai-lseg) [inline packages/elements/src/list/renderer.ts]:
```suggestion
 * Do not use: Renderer is not CSP compliant. 
```

File: packages/elements/src/combo-box/helpers/renderer.ts [hunk 27]
```diff
@@ -1,30 +1,42 @@
...
+    } else if (!value && element.id) {
+      element.removeAttribute('id');
+    }
+
+    return element;
+  };
+};
+
+const deprecationNotice = new DeprecationNotice(
+  'ComboRenderer is deprecated. Use createComboBoxRenderer instead.'
```
Comment (wattachai-lseg) [inline packages/elements/src/combo-box/helpers/renderer.ts]:
```suggestion
  'ComboBoxRenderer is deprecated. Use createComboBoxRenderer instead.'
```

File: packages/elements/src/combo-box/__demo__/index.html (hunk 23 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/combo-box/__demo__/index.html]: Sorry my bad. removed.

**Comment by wattachai-lseg** [comment]
LGTM

**Review by wattachai-lseg — APPROVED** [review]

**Comment by bualoy-napat** [comment]
Verified!

**Pull request merged**